### PR TITLE
Preserve test logs as CI artifacts in case of CI failure.

### PR DESCRIPTION
**Pull request by andy31415**
This makes it easier to debug CI failures.

 #### Problem
CI failures do not include the failure logs, making it hard to investigate failures due to environment.

 #### Summary of Changes
Record failure logs as CI artifacts.

fixes #819

### CODE REVIEW

**Review by gerickson — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]
Thank you! (Waiting on @pan-apple 's review before merging)

**Review by saurabhst — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

**Pull request merged**
